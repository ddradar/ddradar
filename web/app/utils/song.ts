import type { Series, Song } from '@ddradar/core'
import { seriesSet } from '@ddradar/core'

/** LEVEL 1-19 */
export const levels = [...Array(19).keys()].map(i => i + 1)

export const seriesNames = [...seriesSet] as Series[]

/** `0`: DDR 1st, `1`: DDR 2ndMIX, ..., `19`: Dance Dance Revolution WORLD */
export const seriesIndexes = [...Array(seriesSet.size).keys()]

export function getDisplayedBPM({
  minBPM,
  maxBPM,
}: Pick<Song, 'minBPM' | 'maxBPM'>) {
  return minBPM === maxBPM
    ? (`${minBPM}` as const)
    : (`${minBPM}-${maxBPM}` as const)
}

export function shortenSeriesName(series: string) {
  return series.replace(/^(DDR |DanceDanceRevolution )\(?([^)]+)\)?$/, '$2')
}

export const getChartColor = (difficulty: number) =>
  ['blue', 'yellow', 'red', 'green', 'purple'][difficulty]
