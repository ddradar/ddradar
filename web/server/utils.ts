import type { QueryObject } from 'ufo'

/**
 * Get string value from query object.
 * @param query Query object (returned by `useQuery(event)`)
 * @param key Query key
 * @returns Query value (if multiple, returns last value)
 */
export function getQueryString(query: QueryObject, key: string) {
  const value = query[key]
  return Array.isArray(value) ? value[value.length - 1] : value
}

/**
 * Get integer value from query object.
 * @param query Query object (returned by `useQuery(event)`)
 * @param key Query key
 * @returns Integer value or NaN
 */
export function getQueryInteger(query: QueryObject, key: string) {
  const value = getQueryString(query, key)
  if (value === undefined) return NaN
  const num = parseFloat(value)
  return Number.isInteger(num) ? num : NaN
}
