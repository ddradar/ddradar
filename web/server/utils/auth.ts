import type { UserSchema } from '@ddradar/core'
import { fetchLoginUser, fetchUser } from '@ddradar/db'
import type { H3Event } from 'h3'

import { paramsSchema } from '~/schemas/user'

export async function getLoginUserInfo(
  event: H3Event
): Promise<UserSchema | null> {
  const clientPrincipal = getClientPrincipal(event)
  if (!clientPrincipal) return null
  return await fetchLoginUser(clientPrincipal.userId)
}

/**
 * Check user visibility
 * @param event HTTP Event (needs [id] context)
 * @returns UserSchema if user is public or same as login user. otherwise null.
 */
export async function tryFetchUser(event: H3Event): Promise<UserSchema | null> {
  const result = await getValidatedRouterParams(event, paramsSchema.safeParse)
  if (!result.success) return null

  const user = await fetchUser(result.data.id)
  if (!user) return null

  const loginId = getClientPrincipal(event)?.userId ?? ''
  return user.isPublic || user.loginId === loginId ? user : null
}
