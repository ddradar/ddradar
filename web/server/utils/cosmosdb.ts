import { CosmosClient } from '@azure/cosmos'
import type { H3Event } from 'h3'

let _client: CosmosClient | undefined

export function getCosmosClient(event: H3Event): CosmosClient {
  return (_client ??= new CosmosClient(useRuntimeConfig(event).cosmosDBConn))
}
