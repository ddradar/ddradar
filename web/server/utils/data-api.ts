import type { H3Event } from 'h3'

export const $graphql = async <T>(
  event: H3Event,
  query: string,
  variables?: object,
  role?: string
) => {
  const { data } = await $fetch<{ data: T }>('/data-api/graphql', {
    method: 'POST',
    body: { query, variables },
    headers: {
      ...getRequestHeaders(event),
      ...(role ? { 'X-MS-API-ROLE': role } : {}),
    } as Record<string, string>,
  })
  return data
}

type QueryListResult<T> = {
  [key: string]:
    | {
        items: T[]
        hasNextPage: true
        endCursor: string
      }
    | {
        items: T[]
        hasNextPage: false
        endCursor: null
      }
}

export const $graphqlList = async <T>(
  event: H3Event,
  query: string,
  schema: string,
  variables?: object,
  role?: string
) => {
  const result: T[] = []
  let queryResult: QueryListResult<T>[string] = {
    items: [],
    endCursor: null,
    hasNextPage: false,
  }
  do {
    queryResult = (
      await $graphql<QueryListResult<T>>(
        event,
        query,
        {
          ...variables,
          cursor: queryResult?.hasNextPage ? queryResult.endCursor : null,
        },
        role
      )
    )[schema]
    result.push(...queryResult.items)
  } while (queryResult.hasNextPage)

  return result
}
