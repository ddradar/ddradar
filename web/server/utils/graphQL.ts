import { $fetch } from 'ofetch'

/**
 * Call Azure Static Web Apps built-in GraphQL Query requests.
 * https://learn.microsoft.com/azure/static-web-apps/database-overview
 * @param query GraphQL query
 * @param variables GraphQL variables
 * @returns Result data
 */
export function callGraphQL<T>(query: string, variables: object) {
  const headers = useRequestHeaders() as Record<string, string>
  return $fetch<{ data: T }>(`/data-api/graphql`, {
    headers,
    method: 'POST',
    body: { query, variables },
  })
}
