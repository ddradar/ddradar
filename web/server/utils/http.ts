import { createError, sendError } from 'h3'

/**
 * Returns null and send H3Error.
 * @param event HTTP Event
 * @param statusCode HTTP Status code
 * @param message Error message (optional)
 */
export function sendNullWithError(
  event: Parameters<typeof sendError>[0],
  statusCode: number,
  message?: string
) {
  sendError(event, createError({ statusCode, message }))
  return null
}
