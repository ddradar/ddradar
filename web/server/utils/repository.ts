import { CosmosClient } from '@azure/cosmos'
import { SongRepository, UserRepository } from '@ddradar/db'
import type { H3Event } from 'h3'

let _client: CosmosClient | undefined

export function getCosmosClient(event: H3Event): CosmosClient {
  return (_client ??= new CosmosClient(useRuntimeConfig(event).cosmosDBConn))
}

export function getSongRepository(event: H3Event) {
  return new SongRepository(getCosmosClient(event))
}

export function getUserRepository(event: H3Event) {
  return new UserRepository(getCosmosClient(event))
}
