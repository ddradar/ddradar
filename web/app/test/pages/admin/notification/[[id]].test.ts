import { notification } from '@ddradar/core/test/data'
import { mockNuxtImport, mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, test, vi } from 'vitest'

import Page from '~/pages/admin/notification/[[id]].vue'
import { global } from '~/test/utils'

const { useFetchMock, useRouteMock } = vi.hoisted(() => ({
  useFetchMock: vi.fn(),
  useRouteMock: vi.fn(),
}))
mockNuxtImport('useFetch', () => useFetchMock)
mockNuxtImport('useRoute', () => useRouteMock)

describe('/admin/notification', () => {
  test('snapshot test', async () => {
    // Arrange
    /* eslint-disable @typescript-eslint/no-explicit-any */
    vi.mocked(useRoute).mockReturnValue({
      params: { id: notification.id },
    } as any)
    vi.mocked(useFetch).mockReturnValue({
      data: ref(notification),
      execute: vi.fn(),
    } as any)
    /* eslint-enable @typescript-eslint/no-explicit-any */

    // Act
    const wrapper = await mountSuspended(Page, { global })

    // Assert
    expect(wrapper.element).toMatchSnapshot()
  })

  // Computed
  describe('pageTitle', () => {
    test.each([
      [notification.id, 'Update Notification'],
      ['', 'Add Notification'],
    ])('{ id: "%s" } returns "%s"', async (id, title) => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useRoute).mockReturnValue({ params: { id } } as any)
      vi.mocked(useFetch).mockReturnValue({
        data: ref(notification),
        execute: vi.fn(),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountSuspended(Page, { global })

      // Assert
      expect(wrapper.find('h1').text()).toBe(title)
    })
  })
})
