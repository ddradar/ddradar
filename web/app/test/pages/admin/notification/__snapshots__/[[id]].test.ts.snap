// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/admin/notification > snapshot test 1`] = `
<div
  class="flex flex-col lg:grid lg:grid-cols-10 lg:gap-8"
>
  <!--v-if-->
  <div
    class="lg:col-span-10"
  >
    
    <div
      class="relative border-b border-gray-200 dark:border-gray-800 py-8"
    >
      <!--v-if-->
      <div
        class="flex flex-col lg:flex-row items-start gap-6"
      >
        <!--v-if-->
        <div
          class="flex-1"
        >
          <div
            class="flex flex-col lg:flex-row lg:items-center lg:justify-between"
          >
            <h1
              class="text-3xl sm:text-4xl font-bold text-gray-900 dark:text-white tracking-tight"
            >
              
              Update Notification
              
            </h1>
            <!--v-if-->
          </div>
          <!--v-if-->
          
          
        </div>
      </div>
    </div>
    <div
      class="mt-8 pb-24"
    >
      
      <form>
        
        <div
          class=""
        >
          <div
            class=""
          >
            <div
              class="flex content-center items-center justify-between text-sm"
            >
              <label
                class="block font-medium text-gray-700 dark:text-gray-200"
                for="njM8tMUUjQT_1"
              >
                
                Title
                
              </label>
              <!--v-if-->
            </div>
            <!--v-if-->
          </div>
          <div
            class="mt-1 relative"
          >
            
            <div
              class="relative"
            >
              <input
                class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-2.5 py-1.5 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400"
                id="njM8tMUUjQT_1"
                name="title"
                type="text"
                value="新曲を追加しました"
              />
              
              
              <!--v-if-->
              <!--v-if-->
            </div>
            
            <!--v-if-->
          </div>
        </div>
        <div
          class=""
        >
          <div
            class=""
          >
            <div
              class="flex content-center items-center justify-between text-sm"
            >
              <label
                class="block font-medium text-gray-700 dark:text-gray-200"
                for="njM8tMUUjQT_2"
              >
                
                Body
                
              </label>
              <!--v-if-->
            </div>
            <!--v-if-->
          </div>
          <div
            class="mt-1 relative"
          >
            
            <div
              class="relative"
            >
              <textarea
                class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-textarea rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-2.5 py-1.5 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400 resize-none"
                id="njM8tMUUjQT_2"
                name="body"
                rows="3"
                value="下記の譜面情報を追加しました。

- [新曲](/song/foo)"
              />
              
              
            </div>
            
            <!--v-if-->
          </div>
        </div>
        <div
          class=""
        >
          <div
            class=""
          >
            <!--v-if-->
            <!--v-if-->
          </div>
          <div
            class=""
          >
            
            <div
              class="relative flex items-start"
            >
              <div
                class="flex items-center h-5"
              >
                <input
                  class="h-4 w-4 dark:checked:bg-current dark:checked:border-transparent dark:indeterminate:bg-current dark:indeterminate:border-transparent disabled:opacity-50 disabled:cursor-not-allowed focus:ring-0 focus:ring-transparent focus:ring-offset-transparent form-checkbox rounded bg-white dark:bg-gray-900 border border-gray-300 dark:border-gray-700 focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 focus-visible:ring-offset-2 focus-visible:ring-offset-white dark:focus-visible:ring-offset-gray-900 text-primary-500 dark:text-primary-400"
                  id="njM8tMUUjQT_3"
                  name="pinned"
                  type="checkbox"
                />
              </div>
              <div
                class="ms-3 flex flex-col"
              >
                <label
                  class="text-sm font-medium text-gray-700 dark:text-gray-200"
                  for="njM8tMUUjQT_3"
                >
                  
                  ピン留めする
                  
                  <!--v-if-->
                </label>
                <!--v-if-->
              </div>
            </div>
            
            <!--v-if-->
          </div>
        </div>
        <div
          class=""
        >
          <div
            class=""
          >
            <div
              class="flex content-center items-center justify-between text-sm"
            >
              <label
                class="block font-medium text-gray-700 dark:text-gray-200"
                for="njM8tMUUjQT_4"
              >
                
                Type
                
              </label>
              <!--v-if-->
            </div>
            <!--v-if-->
          </div>
          <div
            class="mt-1 relative"
          >
            
            <div
              class="relative"
            >
              <input
                class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-2.5 py-1.5 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400"
                id="njM8tMUUjQT_4"
                name="type"
                type="text"
                value="blue"
              />
              
              
              <!--v-if-->
              <!--v-if-->
            </div>
            
            <!--v-if-->
          </div>
        </div>
        <div
          class=""
        >
          <div
            class=""
          >
            <div
              class="flex content-center items-center justify-between text-sm"
            >
              <label
                class="block font-medium text-gray-700 dark:text-gray-200"
                for="njM8tMUUjQT_5"
              >
                
                Icon
                
              </label>
              <span
                class="text-gray-500 dark:text-gray-400"
              >
                
                
                <a
                  href="https://icon-sets.iconify.design/heroicons/"
                  rel="noopener noreferrer"
                >
                  
                   参考ページ 
                  
                </a>
                
                
              </span>
            </div>
            <!--v-if-->
          </div>
          <div
            class="mt-1 relative"
          >
            
            <div
              class="relative"
            >
              <input
                class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-2.5 py-1.5 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400 ps-9"
                id="njM8tMUUjQT_5"
                name="icon"
                type="text"
                value="i-heroicons-musical-note"
              />
              
              
              <span
                class="absolute inset-y-0 start-0 flex items-center pointer-events-none px-2.5"
              >
                
                <span
                  aria-hidden="true"
                  class="iconify i-heroicons:musical-note flex-shrink-0 text-gray-400 dark:text-gray-500 h-5 w-5"
                />
                
              </span>
              <!--v-if-->
            </div>
            
            <!--v-if-->
          </div>
        </div>
        <button
          class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 px-2.5 py-1.5 shadow-sm text-white dark:text-gray-900 bg-primary-500 hover:bg-primary-600 disabled:bg-primary-500 dark:bg-primary-400 dark:hover:bg-primary-500 dark:disabled:bg-primary-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-primary-500 dark:focus-visible:outline-primary-400 inline-flex items-center"
          type="submit"
        >
          
          
          <!--v-if-->
          
          
          Save
          
          
          <!--v-if-->
          
          
        </button>
        
      </form>
      
    </div>
    
  </div>
  <!--v-if-->
</div>
`;
